packages/scope-manager/src/referencer/Referencer.ts

```diff
@@ -51,23 +67,125 @@ class Referencer extends Visitor {
+              ) {
+                implicitVariablesFromLibs.set(name, TYPE_VALUE);
+                continue;
+              }
+            }
+            // variable is either net-new, or is upgrading to TYPE_VALUE
+            implicitVariablesFromLibs.set(name, variable);
+          }
+        }
+        variablesFromSerializedLibs.set(serialized, implicitVariablesFromLibs);
       }
-      for (const [name, variable] of Object.entries(variables)) {
-        globalScope.defineImplicitVariable(name, variable);
+      variablesFromRawLibs.set(this.#lib, implicitVariablesFromLibs);
+    }
+
+    // Collect all names that are defined or referenced.
+    const allScopes: Set<Scope> = new Set([globalScope]);
+    const allNames: Set<string> = new Set();
+
+    for (const scope of allScopes) {
+      for (const childScope of scope.childScopes) {
+        allScopes.add(childScope);
+      }
+
+      for (const reference of scope.references) {
+        allNames.add(reference.identifier.name);
+      }
+
+      for (const variable of scope.variables) {
+        allNames.add(variable.name);
       }
     }
 
-    // for const assertions (`{} as const` / `<const>{}`)
-    globalScope.defineImplicitVariable('const', {
-      eslintImplicitGlobalSetting: 'readonly',
-      isTypeVariable: true,
-      isValueVariable: false,
-    });
+    // Rules only care about implicit globals if they are referenced or overlap with a local declaration,
+    // so add only those implicit globals.
+    const replacements: Map<Variable, Variable> = new Map();
+    const arraysToProcess: Set<Variable[]> = new Set([globalScope.variables]);
+    const { declaredVariables } = this.scopeManager;
+    for (const name of allNames) {
+      const libVariable: ImplicitLibVariableOptions | undefined =
+        implicitVariablesFromLibs.get(name);
+      if (libVariable) {
+        const existingVariable = globalScope.set.get(name);
+        if (existingVariable) {
+          // This should be cleaned up
+          const newVariable = new ImplicitLibVariable(
+            globalScope,
+            name,
+            libVariable,
+          );
+          replacements.set(existingVariable, newVariable);
+          globalScope.set.set(name, newVariable);
+
+          // Copy over information about the conflicting declaration
+          for (const def of existingVariable.defs) {
+            newVariable.defs.push(def);
+            const parentDeclarations =
+              def.parent && declaredVariables.get(def.parent);
+            const nodeDeclarations = declaredVariables.get(def.node);
+            if (parentDeclarations) {
+              arraysToProcess.add(parentDeclarations);
+            }
+            if (nodeDeclarations) {
+              arraysToProcess.add(nodeDeclarations);
+            }
+          }
+          for (const identifier of existingVariable.identifiers) {
+            newVariable.identifiers.push(identifier);
+          }
```

Comment:
It would be cleaner to give `GlobalScope` the `implicitVariablesFromLibs` map and have it intercept `defineVariable` calls where the variable name is present in the map and define an `ImplicitLibVariable` instead of the variable being requested, then add the extra definition record.

Reply:
Avoids the array reprocessing, but requires a lot of plumbing, so not going to take this approach for now.